project-api/project-api-test/src/test/java/io/ballerina/projects/test/PackageResolutionLocalTests.java

```diff
@@ -0,0 +1,431 @@
+
+        Assert.assertEquals(readFileAsString(projectDirPath.resolve(DEPENDENCIES_TOML)),
+                readFileAsString(projectDirPath.resolve(RESOURCE_DIR_NAME).resolve("Dependencies.toml")));
+
+        // If the user removes the local dependency from the Ballerina.toml
+        projectDirPath = PROJECTS_DIRECTORY.resolve("project_0002_2");
+
+        // Build project_0002:1.0.0
+        buildProject = BuildProject.load(projectEnvironmentBuilder, projectDirPath);
+        buildProject.save();
+        expectedDiagnostics(buildProject, 1, "undefined function 'funcA1'");
+        //failIfDiagnosticsExists(buildProject);
+
+        Assert.assertEquals(readFileAsString(projectDirPath.resolve(DEPENDENCIES_TOML)),
+                readFileAsString(projectDirPath.resolve(RESOURCE_DIR_NAME).resolve("Dependencies.toml")));
+
+        // Delete it manually since aftermethod only works for previous project
+        deleteDependenciesToml(projectDirPath);
+    }
+
+    @Test(description = "Tests Local repository packages (published, but patched as a newer unpublished version) " +
+            "as direct dependencies", dependsOnMethods = "testCase0002")
+    public void testCase0003(ITestContext ctx) throws IOException {
+
+        /*
+            Updated version of package_a to 0.1.1, rebuilt and pushed to local
+            Built project_0003 with existing Dependencies.toml ---> resolved package_a:0.0.2 from central
+            Changed Ballerina.toml of project_0003 to 0.1.1 and build again ---> resolved package_a:0.1.1 from local
+         */
+
+        Path projectDirPath = PROJECTS_DIRECTORY.resolve("project_0003");
+        ctx.getCurrentXmlTest().addParameter("packagePath", String.valueOf(projectDirPath));
+
+        // project_003 --> package_a:0.0.2
+        // Push package_a:0.1.1 to local
+        cacheDependencyToLocalRepository(PACKAGES_DIRECTORY.resolve("package_a_0_1_1"));
+
+        // Check if its actually cached to local properly
+        Path localCache = USER_HOME.resolve(ProjectConstants.REPOSITORIES_DIR)
+                .resolve(LOCAL_REPOSITORY_NAME).resolve(ProjectConstants.BALA_DIR_NAME).resolve(ORG).resolve(
+                        "package_a").resolve("0.1.1").resolve("java11").resolve("dependency-graph.json");
+        Assert.assertTrue(localCache.toFile().exists());
+
+        // package_a should already be pushed to central so it should pick from there
+        // Build project_0003:1.0.0
+        BuildProject buildProject = BuildProject.load(projectEnvironmentBuilder, projectDirPath);
+        buildProject.save();
+        failIfDiagnosticsExists(buildProject);
+
+        Assert.assertEquals(readFileAsString(projectDirPath.resolve(DEPENDENCIES_TOML)),
+                readFileAsString(projectDirPath.resolve(RESOURCE_DIR_NAME).resolve("Dependencies.toml")));
+
+        // Change the Ballerina.toml to include the new version to be pulled from local
+        projectDirPath = PROJECTS_DIRECTORY.resolve("project_0003_2");
```

Comment:
Shall we use the edit APIs to add the repository instead of keeping two projects? This will cove the LS use case where the user would use the VSCode editor to update the file.